### PR TITLE
JENKINS-48737 Adding lightweight checkouts for Bitbucket Pull Requests

### PR DESCRIPTION
**Pull request by benjaminfuchs**
Proposing a fix for https://issues.jenkins-ci.org/browse/JENKINS-48737

Tested on our Bitbucket -> Jenkins setup.

### CODE REVIEW

**Comment by tarioch** [comment]
#114 seems to make more sense than this fix as it fixes the source issue

**Comment by benjaminfuchs** [comment]
@tarioch I totally agree, but this fix will just affect the light weight checkout. So compared to #114 it just fixes this problem and doesn't change the environment variables CHANGE_TARGET.

**Comment by BrunoLavitForgerock** [comment]
FYI, we have applied #117 and #111 on the master branch and the lightweight checkout is now working for our postcommit and PR pipeline jobs.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java [hunk 1]
```diff
@@ -855,12 +855,20 @@ private void collectFileAndDirectories(BitbucketSCMFile parent, List<Map> values
     public InputStream getFileContent(BitbucketSCMFile file) throws IOException, InterruptedException {
         List<String> lines = new ArrayList<>();
         int start=0;
+        String ref;
+        if(file.getRef().matches("PR-\\d+")) {  // because JENKINS-48737
+            String prId = file.getRef().replace("PR-", "");
+            ref = "refs/pull-requests/" + prId + "/from";
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java]: Could respect the pull request merge strategy set on the build configuration

Comment (fengxx) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java]: agree with @casz, could we use `pull-requests/id/merge` as ref for merge strategy?

Comment (casz) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java]: indeed, which I was hinting at 👍

Comment (benjaminfuchs) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java]: Thanks for the feedback. I will pick up your comments. Hopefully I have time this week. :+1:

**Comment by honnel** [comment]
We are affected by this issue, too. Any plans for next release?

**Comment by benjaminfuchs** [comment]
@casz @fengxx Finally I had time reworking this. I changed it, so the checkout strategy is taken into account.

**Comment by benjaminfuchs** [comment]
So I tested it in our environment. It is working but this print is now wrong "Obtained Jenkinsfile from XXXX" since it should mention, that it used the merge ref.
Also if you choose "Merging the pull request with the current target branch revision" it is still using heavy weight checkout and doesn't even try leight weight version. Seems there have to be more changes to support both checkout strategies.

**Comment by casz** [comment]
@benjaminfuchs see #114

**Comment by benjaminfuchs** [comment]
@casz if 114 gets merged I'm also fine with that. Is this done with the next release?

**Comment by bjornmagnusson** [comment]
What is missing for getting this merged?  This is blocker for us.

**Comment by amuniz** [comment]
@tarioch said this issue is also fixed by #114, did someone check that?

**Comment by tarioch** [comment]
As far as I can tell #114 fixes this issue for the "current pull request revision" strategy but it doesn't fix it for the "merge" strategy. Correct me please @benjaminfuchs

**Comment by benjaminfuchs** [comment]
@tarioch Yes #114 will fix the "current pull request revision" but not the "merge" strategy. I think the "merge" strategy is the most common use-case, so I guess this pull request should be merged independent from #114.

**Comment by tarioch** [comment]
@benjaminfuchs I think after #114 is merged the first part

```
if (ref.matches("PR-\\d+")) { // because JENKINS-48737
```

would need to be changed because I think that would then be the branch name of the PR.

**Comment by bjornmagnusson** [comment]
Any updates?

**Comment by casz** [comment]
I'll try and look into this and #114

**Comment by benjaminfuchs** [comment]
@casz Did you plan to finish this? After #114 is merged I would try to have a look in this one this weekend.  Just want sync to avoid double work.

**Comment by casz** [comment]
@benjaminfuchs please look into it 👍

**Comment by bjornmagnusson** [comment]
any ETA on this?

**Comment by casz** [comment]
@benjaminfuchs I checked and PR still does full checkouts.

This needs to adapt: https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/f442c48e26d427f8d1ed14d2a4f976ef7ef018c1/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/filesystem/BitbucketSCMFileSystem.java#L137-L141

**Comment by casz** [comment]
Actually most of the logic can be moved to that spot.

**Comment by casz** [comment]
sadly does not work for cloud

**Comment by benjaminfuchs** [comment]
@casz I guess my changes only work if you set "Discover pull requests from origin" to Strategy "The current pull requets revision"

**Comment by casz** [comment]
I guess we have to check api client if server or cloud for now... until we can find a solution for cloud too

**Comment by casz** [comment]
with the changes from #114 this did indeed never worked at least the current state 😅

**Comment by casz** [comment]
@benjaminfuchs if you could adapt the PR and the tests to:
```java
if (head instanceof BranchSCMHead) {
    ref = head.getName();
} else if (head instanceof PullRequestSCMHead) {
    PullRequestSCMHead pr = (PullRequestSCMHead) head;
    if (apiClient instanceof BitbucketCloudApiClient) {
        if (!(pr.getCheckoutStrategy() == ChangeRequestCheckoutStrategy.MERGE) && pr.getRepository() != null) {
            return new BitbucketSCMFileSystem(apiClient, pr.getOriginName(), rev);
        }
        return null; // TODO support merge revisions somehow for cloud
    }
    String checkoutStrat = "from";
    if (pr.getCheckoutStrategy() == ChangeRequestCheckoutStrategy.MERGE) {
        checkoutStrat = "merge";
    }
    ref = "refs/pull-requests/" + pr.getId() + "/" + checkoutStrat;
} else if (head instanceof BitbucketTagSCMHead) {
     ref = "tags/" + head.getName();
} else {
    return null;
}
```

I would appreciate it 😇🙌

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/filesystem/BitbucketSCMFile.java [hunk 2]
```diff
@@ -119,4 +129,51 @@ protected Type type() throws IOException, InterruptedException {
...
+			throw new UnsupportedOperationException(
+					"Can not get reference for this pull request since checkout strategy can not be determined");
+		}
+		return "refs/pull-requests/" + prId + "/" + source;
+	}
+
+	@CheckForNull
+	private ChangeRequestCheckoutStrategy getCheckoutStrategy() throws IOException, InterruptedException {
+		if (this.isDirectory()) {
+			SCMRevision revision = fileSystem.getRevision();
```
Comment (darxriggs) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/filesystem/BitbucketSCMFile.java]: Having a look at the constructors in [line 76](https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/117/files#diff-f62992a8799c0de0f059c675c93e7708R76) and [line 80](https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/117/files#diff-f62992a8799c0de0f059c675c93e7708R80) `fileSystem` could be `null`.
I would also mark this variable `final` and maybe `@CheckForNull`.

**Comment by bjornmagnusson** [comment]
Any ETA for this very important feature?

**Comment by casz** [comment]
@bjornmagnusson if you feel it is important, feel free to rework the PR.
I'll leave this here.
https://snarky.ca/setting-expectations-for-open-source-participation/

**Comment by casz** [comment]
Superseded by #174

**Pull request closed without merging**